src/v0/destinations/tiktok_ads/transformV2.js

```diff
@@ -0,0 +1,740 @@
+/* eslint-disable camelcase */
+/* eslint-disable @typescript-eslint/naming-convention */
+const set = require('set-value');
+const { ConfigurationError, InstrumentationError } = require('@rudderstack/integrations-lib');
+const { EventType } = require('../../../constants');
+const {
+  constructPayload,
+  defaultRequestConfig,
+  defaultPostRequestConfig,
+  removeUndefinedAndNullValues,
+  defaultBatchRequestConfig,
+  getSuccessRespEvents,
+  isDefinedAndNotNullAndNotEmpty,
+  getDestinationExternalID,
+  getHashFromArrayWithDuplicate,
+  checkInvalidRtTfEvents,
+  handleRtTfSingleEventError,
+} = require('../../util');
+const { getContents, hashUserField } = require('./util');
+const {
+  trackMappingV2,
+  trackEndpointV2,
+  eventNameMapping,
+  PARTNER_NAME,
+  maxBatchSizeV2,
+} = require('./config');
+const { JSON_MIME_TYPE } = require('../../util/constant');
+
+/**
+ * Generated track payload for an event using TikTokTrackV2.json config mapping,
+ * hashing user properties and
+ * defining contents from products
+ * @param {*} message
+ * @param {*} Config
+ * @param {*} event
+ * @returns track payload
+ */
+const getTrackResponsePayload = (message, Config, event) => {
+  const payload = constructPayload(message, trackMappingV2);
+
+  // if contents is not an array converting it into array
+  if (payload.properties?.contents && !Array.isArray(payload.properties.contents)) {
+    payload.properties.contents = [payload.properties.contents];
+  }
+
+  // if contents is not present but we have properties.products present which has fields with superset of contents fields
+  if (payload.properties && !payload.properties?.contents && message.properties?.products) {
+    // retreiving data from products only when contents is not present
+    payload.properties.contents = getContents(message, false);
+  }
+
+  // getting externalId and hashing it and storing it in
+  const externalId = getDestinationExternalID(message, 'tiktokExternalId');
+  if (isDefinedAndNotNullAndNotEmpty(externalId)) {
+    set(payload, 'user.external_id', externalId);
+  }
+  if (Config.hashUserProperties && isDefinedAndNotNullAndNotEmpty(payload.user)) {
+    payload.user = hashUserField(payload.user);
+  }
+  payload.event = event;
+  // add partner name and return payload
+  return removeUndefinedAndNullValues(payload);
+};
+
+const trackResponseBuilder = async (message, { Config }) => {
+  const { eventsToStandard, sendCustomEvents, accessToken, pixelCode } = Config;
+
+  let event = message.event?.toLowerCase().trim();
+  if (!event) {
+    throw new InstrumentationError('Event name is required');
+  }
+
+  const standardEventsMap = getHashFromArrayWithDuplicate(eventsToStandard);
+
+  if (!sendCustomEvents && eventNameMapping[event] === undefined && !standardEventsMap[event]) {
```

Comment:
Can't we stop the use of sendCustomEvents? As this is v2 by default we send the event as a custom event when it is not mapped.

Reply:
Yeah we can do it. @krishna2020 is it good if we automatically custom events in events 2.0 ?

Reply:
After discussing it with @krishna2020 we are going to let it be there to let customers decide if they want to send custom events or not.
No change in functionality basically